panoply/datasource.py

```diff
@@ -18,13 +24,58 @@ def log(self, *msgs):
         if "logger" in self.options:
             self.options["logger"](msgs)
         else:
-            print msgs
+            print(msgs)
 
-    def progress(self, loaded, total, msg):
+    def progress(self, loaded, total, msg=''):
         self.fire("progress", {
             "loaded": loaded,
             "total": total,
             "msg": msg
         })
 
+# a decorator used to invalidate the access_token for oauth based data sources
+# this should be used on every method in the data source that fetches data
+# from the server (controlled by this oauth), and that needs to be invalidated
+def invalidate_token(refresh_url, callback=None,
```

Comment:
Changing the source object is not a side effect - this is exactly the purpose of this function. What's the point of having a function the calls a URL (the refresh url), if it is not going to do anything with it and each DS would have to do the all the "hard" work (instead of catching an exception, it would have to explicitly check if the token is valid and if not, invoke the function and afterwards make sure that the source is updated.
Let alone that you would have to make sure that each DS that uses tokens, saves them to the source